File: brainfuck.py
#/usr/bin/env python

import sys
from collections import defaultdict

def interpret(program):
    instrPtr = 0
    mem = [0]
    ptr = 0
    d = defaultdict(int, {'[':1, ']':-1})
    while(instrPtr < len(program)):
        instr = program[instrPtr]
        if instr == '>':
            ptr += 1
            if ptr >= len(mem):
                mem.append(0)
        elif instr == '<':
            ptr -= 1
        elif instr == '+':
            mem[ptr] += 1
        elif instr == '-':
            mem[ptr] -= 1
        elif instr == '.':
            sys.stdout.write(chr(mem[ptr]))
        elif instr == ',':
            mem[ptr] = ord(sys.stdin.read(1))
        elif instr == '[' and mem[ptr] == 0 or instr == ']' and mem[ptr] != 0:
            nest = 1
            while(nest != 0):
                instrPtr += d[instr]
                nest += d[instr] * d[program[instrPtr]]
        instrPtr += 1

interpret(open(sys.argv[1]).read())

File: brainfuck.c
#include <stdio.h>

int d(char c){
    return '\\' - c; // ASCII hack
}
void interpret(char *program){
    char *progPtr = program;
    char mem[10000];
    char *ptr = mem;
    while(*progPtr != '\0'){
        switch (*progPtr){
            case '>': ++ptr; break;
            case '<': --ptr; break;
            case '+': ++*ptr; break;
            case '-': --*ptr; break;
            case '.': putchar(*ptr); break;
            case ',': *ptr = getchar(); break;
            case '[':
            case ']':
                if (*progPtr == '[' && *ptr == 0 || *progPtr == ']' && *ptr != 0){
                    int nest = 1;
                    int instr = d(*progPtr);
                    while (progPtr && nest != 0){
                        progPtr += instr;
                        if (*progPtr == '[' || *progPtr == ']'){
                            nest += d(*progPtr) * instr;
                        }
                    }
                }
                break;
            default:
                break;
        }
        progPtr++;
    }
}

int main(int argc, char** argv){
    if ( argc != 2 ){
        printf( "usage: %s filename", argv[0] );
        return 1;
    }
    FILE *fp = fopen(argv[1], "rb");
    fseek(fp, 0, SEEK_END);
    long siz = ftell(fp);
    rewind(fp);
    char buffer[siz + 1];
    fread(buffer, 1, siz, fp);
    buffer[siz] = '\0';
    interpret(buffer);
    return 0;
}

